utils/utils.c

```diff
@@ -845,3 +849,64 @@ int AlmostEqualRelativeAndAbs_double(double A, double B,
 }
 
 /* #undef __USE_XOPEN2K */
+
+/* A parallel cumulative sum
+   Output convention is: cumsum[0] = 0; cumsum[N-1] = sum(a[0:N-1]);
+   The algorithm is:
+   - Divide the array into `nthreads` chunks
+   - cumsum within each chunk
+   - compute the "offset" for each chunk by summing the cumsum at the tail of all previous chunks
+   - apply the offset
+*/
+void parallel_cumsum(const int64_t *a, const int64_t N, int64_t *cumsum){
+    #ifdef _OPENMP
+    int nthreads = omp_get_max_threads();
+    #else
+    int nthreads = 1;
+    #endif
+    
+    int64_t min_N_per_thread = 10000;  // heuristic to avoid overheads
+    if(N/min_N_per_thread < nthreads){
+        nthreads = N/min_N_per_thread;
+    }
+    if(nthreads < 1){
+        nthreads = 1;
+    }
+    
+    #ifdef _OPENMP
+    #pragma omp parallel num_threads(nthreads)
+    #endif
+    {
+        #ifdef _OPENMP
+        int tid = omp_get_thread_num();
+        #else
+        int tid = 0;
+        #endif
+        
+        int64_t cstart = N*tid/nthreads;
+        int64_t cend = N*(tid+1)/nthreads;
+        cumsum[cstart] = cstart > 0 ? a[cstart-1] : 0;
+        for(int64_t c = cstart+1; c < cend; c++){
+            cumsum[c] = a[c-1] + cumsum[c-1];
+        }
+        
+        #ifdef _OPENMP
+        #pragma omp barrier
+        #endif
+        
+        int64_t offset = 0;
+        for(int t = 0; t < tid; t++){
+            offset += cumsum[N*(t+1)/nthreads-1];
+        }
```

Comment:
Could this be annotated with an `#pragma omp simd for reduction(+:offset)`?

Reply:
Although this is probably an immeasurable amount of time :D

Reply:
Yeah, seems hard to gain much here.  I bet that would be safe with the reduction clause, but I also kind of doubt it would do a SIMD transform, given that the memory accesses are non-contiguous.